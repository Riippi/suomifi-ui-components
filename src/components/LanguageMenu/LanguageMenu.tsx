import React, { Component, ReactNode, Fragment } from 'react';
import classnames from 'classnames';
import { HtmlSpan } from '../../reset/HtmlSpan/HtmlSpan';
import {
  Menu,
  MenuButton,
  MenuList,
  MenuListProps,
  MenuItem,
  MenuLink,
  MenuPopover,
  MenuPopoverProps,
} from '@reach/menu-button';
import { logger } from '../../utils/logger';
import { Omit } from '../../utils/typescript';

export {
  MenuList,
  MenuListProps,
  MenuItem,
  MenuLink,
  MenuPopover,
  MenuPopoverProps,
};

const baseClassName = 'fi-language-menu';
export interface LanguageMenuItemProps {
  /** Operation to run on select */
  onSelect: () => void;
  /** Item content */
  children: ReactNode;
}

type SupportedMenuLinkComponent = keyof JSX.IntrinsicElements;

interface LanguageMenuLinkPropsWithType {
  type: 'menulink';
  /** Url to direct to */
  href: string;
  /** Item content */
  children: ReactNode;
  className?: string;
  as?: SupportedMenuLinkComponent;
}

export interface LanguageMenuLinkProps
  extends Omit<LanguageMenuLinkPropsWithType, 'type'> {}

export type LanguageMenuListItemsProps =
  | LanguageMenuItemProps
  | LanguageMenuLinkPropsWithType;
type OptionalLanguageMenuListProps = {
  [K in keyof MenuListProps]?: MenuListProps[K];
};

export interface LanguageMenuProps {
  /** Name or content of menubutton */
  name: React.ReactNode | ((props: { isOpen: boolean }) => React.ReactNode);
  /** Custom classname to extend or customize */
  className?: string;
  /** Custom classname to extend or customize */
  languageMenuButtonClassName?: string;
  /** Custom classname to apply when menu is open */
  languageMenuOpenButtonClassName?: string;
  /** Properties given to LanguageMenu's List-component, className etc. */
  languageMenuListProps?: OptionalLanguageMenuListProps;
  languageMenuListComponent?: React.ComponentType<
    OptionalLanguageMenuListProps
  >;
  /** Menu items: MenuItem or MenuLink */
  children?: Array<React.ReactElement<LanguageMenuListItemsProps>>;
}

export class LanguageMenu extends Component<LanguageMenuProps> {
  render() {
    const {
      children,
      name,
      className,
      languageMenuButtonClassName: menuButtonClassName,
      languageMenuOpenButtonClassName: menuButtonOpenClassName,
      languageMenuListProps: menuListProps = {},
      languageMenuListComponent: MenuListComponentReplace,
      ...passProps
    } = this.props;

    if (React.Children.count(children) < 1) {
      logger.warn(`Menu '${name}' does not contain items`);
      return null;
    }
    return (
      <HtmlSpan className={classnames(className, baseClassName)}>
        <Menu>
          {({ isOpen }: { isOpen: boolean }) => {
            return (
              <Fragment>
                <MenuButton
                  {...passProps}
                  className={classnames(
                    menuButtonClassName,
                    isOpen && menuButtonOpenClassName,
                  )}
                >
                  {name}
                </MenuButton>
                {!!MenuListComponentReplace ? (
                  <MenuListComponentReplace {...menuListProps}>
                    {children}
                  </MenuListComponentReplace>
                ) : (
                  <MenuList {...menuListProps}>{children}</MenuList>
                )}
              </Fragment>
            );
          }}
        </Menu>
      </HtmlSpan>
    );
  }
}
