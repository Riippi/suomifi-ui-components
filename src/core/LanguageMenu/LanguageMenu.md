<!-- ```js
import { LanguageMenu, LanguageMenuItem, LanguageMenuLink } from 'suomifi-ui-components';

<LanguageMenu className="language-menu-test" name="LanguageMenu">
  <LanguageMenuItem onSelect={() => console.log('LanguageMenu test 1')}>
    Item 1
  </LanguageMenuItem>
  <LanguageMenuItem onSelect={() => console.log('LanguageMenu test 2')}>
    Item 2
  </LanguageMenuItem>
  <LanguageMenuLink href="http://www.suomi.fi/">Suomi.fi</LanguageMenuLink>
</LanguageMenu>;
``` -->

```js
import {
  LanguageMenu,
  LanguageMenuItem,
  LanguageMenuLink
} from 'suomifi-ui-components';

<LanguageMenu className="language-menu-language-test" name="FI">
  <LanguageMenuItem onSelect={() => console.log('FI')} selected>
    Suomeksi (FI)
  </LanguageMenuItem>
  <LanguageMenuLink href="/sv">PÃ¥ svenska (SV)</LanguageMenuLink>
</LanguageMenu>;
```
