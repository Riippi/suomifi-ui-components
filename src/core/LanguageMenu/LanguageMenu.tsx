import React, { Component, ReactNode, Fragment } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { classnamesValue } from '../../utils/typescript';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp, InternalTokensProp } from '../theme';
import {
  baseStyles,
  languageMenuPopoverStyles,
} from './LanguageMenu.baseStyles';
import {
  LanguageMenu as CompLanguageMenu,
  LanguageMenuProps as CompLanguageMenuProps,
  LanguageMenuPopoverItemsProps,
  MenuItems as CompMenuItems,
  MenuPopover as CompMenuPopover,
  MenuPopoverProps as CompMenuPopoverProps,
  PRect,
} from '../../components/LanguageMenu/LanguageMenu';
import {
  LanguageMenuItemLanguage,
  LanguageMenuItemLanguageProps,
  LanguageMenuLinkLanguage,
  LanguageMenuLinkLanguageProps,
} from './LanguageMenuItem';

import { Icon } from '../Icon/Icon';

const itemClassName = 'fi-language-menu_item';
const itemLangClassName = 'fi-language-menu-language_item';
const buttonClassName = 'fi-language-menu_button';
const buttonOpenClassName = 'fi-language-menu-language_button_open';
const buttonLangClassName = 'fi-language-menu-language_button';

const popoverClassName = 'fi-language-menu_popover';
const popoverLangClassName = 'fi-language-menu-language_popover';
const iconLangClassName = 'fi-language-menu-language_icon';

export interface LanguageMenuProps extends CompLanguageMenuProps, TokensProp {}

const StyledLanguageMenu = styled(
  ({ tokens, ...passProps }: LanguageMenuProps & InternalTokensProp) => (
    <CompLanguageMenu {...passProps} />
  ),
)`
  ${(props) => baseStyles(props)}
`;

const LanguageMenuPopoverWithProps = (
  children: ReactNode,
  addClass?: classnamesValue,
) =>
  React.Children.map(
    children,
    (child: React.ReactElement<LanguageMenuPopoverItemsProps>) => {
      // Set defaul component-prop/type to be 'a' needed for links
      if (React.isValidElement(child)) {
        return React.cloneElement(child, {
          as: 'a',
          className: classnames(itemClassName, addClass),
        });
      }
      return child;
    },
  );

const languageName = (name: ReactNode) => (
  <Fragment>
    {name}
    <Icon icon="chevronDown" className={iconLangClassName} />
  </Fragment>
);

interface LanguageMenuPopoverProps extends CompMenuPopoverProps, TokensProp {}

const LanguageMenuPopoverPosition = (
  targetRect: PRect | null,
  popoverRect: PRect | null,
): React.CSSProperties => {
  if (!targetRect || !popoverRect) {
    return {};
  }
  return {
    left: `${targetRect.left - popoverRect.width + targetRect.width}px`,
    // eslint-disable-next-line no-undef
    top: `${targetRect.top + targetRect.height + window.pageYOffset}px`,
    maxWidth: `${Math.max(
      targetRect.width,
      targetRect.width + targetRect.left - 20,
    )}px`,
    minWidth: `${targetRect.width - 2}px`,
  };
};

const StyledMenuPopover = styled(
  ({ tokens, children, ...passProps }: LanguageMenuPopoverProps) => (
    <CompMenuPopover {...passProps} position={LanguageMenuPopoverPosition}>
      <CompMenuItems>{children}</CompMenuItems>
    </CompMenuPopover>
  ),
)`
  ${(props) => languageMenuPopoverStyles(props.theme)}
`;

class LanguageMenuVariation extends Component<LanguageMenuProps> {
  render() {
    const {
      children,
      name,
      className,
      languageMenuPopoverComponent: MenuPopoverComponentProp,
      ...passProps
    } = withSuomifiDefaultProps(this.props);
    const languageMenuButtonClassName = classnames(
      buttonClassName,
      buttonLangClassName,
      className,
    );
    const menuPopoverProps = {
      className: classnames(popoverClassName, popoverLangClassName),
    };

    return (
      <Fragment>
        <StyledLanguageMenu
          {...passProps}
          name={languageName(name)}
          languageMenuButtonClassName={languageMenuButtonClassName}
          languageMenuOpenButtonClassName={buttonOpenClassName}
          languageMenuPopoverProps={menuPopoverProps}
          languageMenuPopoverComponent={
            !!MenuPopoverComponentProp
              ? MenuPopoverComponentProp
              : StyledMenuPopover
          }
        >
          {LanguageMenuPopoverWithProps(children, itemLangClassName)}
        </StyledLanguageMenu>
      </Fragment>
    );
  }
}

/**
 * <i class="semantics" />
 * Use for dropdown menu.
 */
export class LanguageMenu extends Component<LanguageMenuProps> {
  static languageItem = (props: LanguageMenuItemLanguageProps) => {
    return <LanguageMenuItemLanguage {...props} />;
  };

  static LinkLanguage = (props: LanguageMenuLinkLanguageProps) => {
    return <LanguageMenuLinkLanguage {...props} />;
  };

  render() {
    return <LanguageMenuVariation {...this.props} />;
  }
}
