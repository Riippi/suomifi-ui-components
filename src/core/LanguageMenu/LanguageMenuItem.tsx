import React, { Component } from 'react';
import {
  LanguageMenuItem as CompLanguageMenuItem,
  LanguageMenuLink as CompLanguageMenuLink,
  LanguageMenuItemProps,
  LanguageMenuLinkProps,
} from '../../components/LanguageMenu/LanguageMenu';
import {
  LanguageMenuItemLanguage,
  LanguageMenuItemLanguageProps,
} from './LanguageMenuItemLanguage';
import {
  LanguageMenuLinkLanguage,
  LanguageMenuLinkLanguageProps,
} from './LanguageMenuLinkLanguage';
export {
  LanguageMenuItemProps,
  LanguageMenuLinkProps,
  LanguageMenuItemLanguage,
  LanguageMenuItemLanguageProps,
  LanguageMenuLinkLanguage,
  LanguageMenuLinkLanguageProps,
};

export class LanguageMenuItem extends Component<LanguageMenuItemProps> {
  static language = (props: LanguageMenuItemLanguageProps) => {
    return <LanguageMenuItemLanguage {...props} />;
  };

  render() {
    return CompLanguageMenuItem;
  }
}

export class LanguageMenuLink extends Component<LanguageMenuLinkProps> {
  static language = (props: LanguageMenuLinkLanguageProps) => {
    return <LanguageMenuLinkLanguage {...props} />;
  };

  render() {
    return CompLanguageMenuLink;
  }
}
