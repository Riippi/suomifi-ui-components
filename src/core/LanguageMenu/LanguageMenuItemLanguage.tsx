import React from 'react';
import classnames from 'classnames';
import {
  LanguageMenuItem as CompLanguageMenuItem,
  LanguageMenuItemProps,
} from '../../components/LanguageMenu/LanguageMenu';

export interface LanguageMenuItemLanguageProps extends LanguageMenuItemProps {
  /** Show item as selected one */
  selected?: boolean;
  className?: string;
}

export const LanguageMenuItemLanguage = ({
  selected,
  className,
  ...passProps
}: LanguageMenuItemLanguageProps) => (
  <CompLanguageMenuItem
    {...passProps}
    className={classnames(className, {
      'fi-language-menu-lang-item-selected': selected,
    })}
  />
);
