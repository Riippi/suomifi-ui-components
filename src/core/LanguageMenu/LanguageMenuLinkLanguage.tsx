import React from 'react';
import classnames from 'classnames';
import {
  MenuLink as CompMenuLink,
  LanguageMenuLinkProps,
} from '../../components/LanguageMenu/LanguageMenu';

export interface LanguageMenuLinkLanguageProps extends LanguageMenuLinkProps {
  /** Show item as selected one */
  selected?: boolean;
  className?: string;
}

export const LanguageMenuLinkLanguage = ({
  selected,
  className,
  ...passProps
}: LanguageMenuLinkLanguageProps) => (
  <CompMenuLink
    {...passProps}
    className={classnames(className, {
      'fi-language-menu-lang-item-selected': selected,
    })}
  />
);
