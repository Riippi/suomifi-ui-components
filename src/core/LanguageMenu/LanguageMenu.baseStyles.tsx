import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../theme';
import { LanguageMenuProps } from './LanguageMenu';
import { element, focus } from '../theme/reset';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme & Partial<LanguageMenuProps>) => css`
  & > [data-reach-menu-button].fi-language-menu_button {
    ${element({ theme })}
    ${theme.typography.bodyText}
    ${focus({ theme })}
    cursor: pointer;
    &.fi-language-menu-language_button {
      ${element({ theme })}
      ${theme.typography.actionElementInnerTextBold}
      padding: 5px ${theme.spacing.xs} 5px ${theme.spacing.s};
      line-height: 28px;
      background-color: ${theme.colors.whiteBase};
      border: 1px solid ${theme.colors.depthBase};
      border-radius: ${theme.radius.basic};
      text-transform: uppercase;
      & > .fi-language-menu-language_icon {
        height: 1.2em;
        width: 1.2em;
        transform: translateY(0.3em); 
        margin-left: ${theme.spacing.xs};
      }
    }
    &.fi-language-menu-language_button_open {
        & > .fi-language-menu-language_icon.fi-language-menu-language_icon {
          transform: translateY(0.2em) rotate(180deg);
        }
      }
  }
`,
);

export const languageMenuPopoverStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme) => css`
  &[data-reach-menu-popover].fi-language-menu_popover {
    ${element({ theme })}
    ${theme.typography.bodyText}
    margin-top: -2px;
    background-color: ${theme.colors.whiteBase};
    border: none;
    box-shadow: ${theme.shadows.menuShadow};
    &.fi-language-menu-language_popover {
      ${theme.typography.actionElementInnerText}
      position: absolute;
      box-sizing: content-box;
      margin-top: 12px;
      padding: 10px 0;
      border: 1px solid ${theme.colors.depthBase};
      border-radius: ${theme.radius.basic};
      &:before,
      &:after {
        content: '';
        position: absolute;
        height: 0;
        width: 0;
        bottom: 100%;
        right: 20px;
        border: solid transparent;
        pointer-events: none;
      }
      &:before {
        border-bottom-color: ${theme.colors.depthBase};
        border-width: 8px;
        margin-right: -8px;
      }
      &:after {
        border-bottom-color: ${theme.colors.whiteBase};
        border-width: 7px;
        margin-right: -7px;
      }
    }
  }

  & [data-reach-menu-items] {
    border: 0;
    padding: 0;
  }

  & [data-reach-menu-item].fi-language-menu_item {
    ${element({ theme })}
    ${theme.typography.bodyText}
    &[data-selected] {
      ${theme.typography.bodyText}
      color: ${theme.colors.blackBase};
      background-color: ${theme.colors.highlightLight3};
    }
    &.fi-language-menu-language_item,
    &[data-selected].fi-language-menu-language_item {
      ${theme.typography.actionElementInnerText}
      padding: 6px 20px 6px 14px;
      border-left: 6px solid transparent;
      background-color: transparent;
      &.fi-language-menu-lang-item-selected {
        ${theme.typography.actionElementInnerTextBold};
      }
    }
    &[data-selected].fi-language-menu-language_item {
      border-left-color: ${theme.colors.highlightBase};
    }
    &:focus {
      outline: 0;
    }
  }
`,
);
