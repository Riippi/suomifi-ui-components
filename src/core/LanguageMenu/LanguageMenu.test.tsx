import React from 'react';
import { axeTest } from '../../utils/test/axe';

import { LanguageMenu } from './LanguageMenu';
import { LanguageMenuItem, LanguageMenuLink } from './LanguageMenuItem';

const doNothing = () => ({});

const TestMenuLanguage = (
  <LanguageMenu className="menu-language-test" name="FI">
    <LanguageMenuItem onSelect={doNothing} selected={true}>
      Suomeksi (FI)
    </LanguageMenuItem>
    <LanguageMenuLink href="/sv">PÃ¥ svenska (SV)</LanguageMenuLink>
  </LanguageMenu>
);

test('should not have basic accessibility issues', axeTest(TestMenuLanguage));
