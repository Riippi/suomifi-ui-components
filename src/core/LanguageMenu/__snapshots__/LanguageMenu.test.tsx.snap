// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  display: inline-block;
  vertical-align: baseline;
}

.c1 > [data-reach-menu-button].fi-language-menu_button {
  color: hsl(0,0%,16%);
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  cursor: pointer;
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button {
  color: hsl(0,0%,16%);
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  padding: 9px 10px;
  line-height: 24px;
  background-color: hsl(0,0%,100%);
  border: 1px solid transparent;
  border-radius: 2px;
  word-break: break-word;
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button > .fi-language-menu-language_icon {
  height: 1em;
  width: 1em;
  -webkit-transform: translateY(0.2em);
  -ms-transform: translateY(0.2em);
  transform: translateY(0.2em);
  margin-left: 10px;
  fill: hsl(212,63%,45%);
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button:focus {
  outline: 0;
  position: relative;
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button:hover {
  border-color: hsl(202,7%,80%);
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button_open {
  border-color: hsl(202,7%,80%);
}

.c1 > [data-reach-menu-button].fi-language-menu_button.fi-language-menu-language_button_open > .fi-language-menu-language_icon.fi-language-menu-language_icon {
  -webkit-transform: translateY(0.2em) rotate(180deg);
  -ms-transform: translateY(0.2em) rotate(180deg);
  transform: translateY(0.2em) rotate(180deg);
}

<div>
  <span
    class="c0 c1 fi-language-menu"
  >
    <button
      aria-controls="menu--1"
      aria-haspopup="true"
      class="fi-language-menu_button fi-language-menu-language_button menu-language-test"
      data-reach-menu-button=""
      id="menu-button--menu--1"
      type="button"
    >
      FI
      <svg
        aria-hidden="true"
        class="fi-icon c2 fi-language-menu-language_icon"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M13 15.6c-.3.3-.6.4-1 .4s-.7-.1-1-.4l-5.6-5.3c-.5-.5-.5-1.4 0-1.9s1.4-.5 2 0l4.6 4.4 4.6-4.4c.5-.5 1.4-.5 2 0 .5.5.5 1.4 0 1.9L13 15.6z"
        />
      </svg>
    </button>
  </span>
</div>
`;
